pkg/cmd/server/origin/master.go

```diff
@@ -2,214 +2,145 @@ package origin
+	}
+	if c.Options.OAuthConfig != nil {
+		ret.MasterPublicURL = c.Options.OAuthConfig.MasterPublicURL
+	}
+	if c.Options.TemplateServiceBrokerConfig != nil {
+		ret.TemplateNamespaces = c.Options.TemplateServiceBrokerConfig.TemplateNamespaces
+	}
 
-var excludedV1Types = sets.NewString()
+	return ret
+}
 
 // Run launches the OpenShift master by creating a kubernetes master, installing
 // OpenShift APIs into it and then running it.
-func (c *MasterConfig) Run(kc *kubernetes.MasterConfig, assetConfig *AssetConfig, stopCh <-chan struct{}) {
-	var (
-		messages []string
-		err      error
-	)
-	kc.Master.GenericConfig.BuildHandlerChainFunc, messages, err = c.buildHandlerChain(assetConfig)
+func (c *MasterConfig) Run(kubeAPIServerConfig *kubeapiserver.Config, assetConfig *AssetConfig, stopCh <-chan struct{}) {
+	openshiftNonAPIConfig := c.newOpenshiftNonAPIConfig(*kubeAPIServerConfig.GenericConfig)
+	openshiftNonAPIServer, err := openshiftNonAPIConfig.Complete().New(apiserver.EmptyDelegate, stopCh)
 	if err != nil {
 		glog.Fatalf("Failed to launch master: %v", err)
 	}
 
-	kmaster, err := kc.Master.Complete().New(apiserver.EmptyDelegate)
+	openshiftAPIServerConfig, err := c.newOpenshiftAPIConfig(*kubeAPIServerConfig.GenericConfig)
 	if err != nil {
 		glog.Fatalf("Failed to launch master: %v", err)
 	}
-
-	c.InstallProtectedAPI(kmaster.GenericAPIServer)
-	messages = append(messages, c.kubernetesAPIMessages(kc)...)
-
-	for _, s := range messages {
-		glog.Infof(s, c.Options.ServingInfo.BindAddress)
+	// TODO this is eventually where we end up, with the openshift server completely discrete from the kube one
+	// but this only works *AFTER* we commit to the aggregator.  Right now the aggregator is optional, so we have
+	// to install ourselves in the kubeapiserver
+	// openshiftAPIServer, err := openshiftAPIServerConfig.Complete().New(openshiftNonAPIServer.GenericAPIServer, stopCh)
+	// if err != nil {
+	// 	glog.Fatalf("Failed to launch master: %v", err)
+	// }
+	// // this sets up the openapi endpoints
+	// preparedOpenshiftAPIServer := openshiftAPIServer.GenericAPIServer.PrepareRun()
+
+	// TODO move out of this function to somewhere we build the kubeAPIServerConfig
+	kubeAPIServerConfig.GenericConfig.BuildHandlerChainFunc, err = c.buildHandlerChain(assetConfig)
+	if err != nil {
+		glog.Fatalf("Failed to launch master: %v", err)
 	}
-
-	apiserver := kmaster.GenericAPIServer.PrepareRun()
+	kubeAPIServer, err := kubeAPIServerConfig.Complete().New(openshiftNonAPIServer.GenericAPIServer)
+	if err != nil {
+		glog.Fatalf("Failed to launch master: %v", err)
+	}
+	// TODO this goes away in 3.7 after we commit to the aggregator always being on (even if its just in local mode).
+	// this is installing the openshift APIs into the kubeapiserver
+	installAPIs(openshiftAPIServerConfig, kubeAPIServer.GenericAPIServer)
+	// this sets up the openapi endpoints
+	preparedKubeAPIServer := kubeAPIServer.GenericAPIServer.PrepareRun()
```

Comment:
don't like that we call `PrepareRun()` only here and not for the others. Is it the same in kube-apiserver?